op-node/rollup/derive/engine_controller.go

```diff
@@ -358,10 +348,27 @@ func (e *EngineController) InsertUnsafePayload(ctx context.Context, envelope *et
 		FinalizedBlockHash: e.finalizedHead.Hash,
 	}
 	if e.syncStatus == syncStatusFinishedELButNotFinalized {
-		fc.SafeBlockHash = envelope.ExecutionPayload.BlockHash
-		fc.FinalizedBlockHash = envelope.ExecutionPayload.BlockHash
-		e.SetSafeHead(ref)
-		e.SetFinalizedHead(ref)
+		// Find the highest finalized block by traversing L2 blocks from the tip and
```

Comment:
This finalization procedure does not work for EL clients that do not have the full state available at the L2 block that you end up at (i.e. op-geth during snap sync).

op-node is only able to finalize from blocks where the full state is available. We may want to do a soft finality delay rather than immediately finalizing, but we have to assume that during our first sync that the unsafe head we receive is valid. That's why we only EL sync once, to limit the amount of trust we place in the unsafe head - it only matters on the first sync (similar to subjective safety/checkpoint sync) and then on future restarts of the op-node, we always finalize against L1